File: SamSafeCSharp/wwwroot/original source sam-safe/safe.js
/******************************************************************************************* 
 * The MIT License (MIT)
 * -----------------------------------------------------------------------------------------
 * Copyright (c) 2016 Convergence Modeling LLC
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of 
 * this software and associated documentation files (the "Software"), to deal in the 
 * Software without restriction, including without limitation the rights to use, copy, 
 * modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, 
 * and to permit persons to whom the Software is furnished to do so, subject to the 
 * following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all 
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR 
 * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE 
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, 
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE 
 * USE OR OTHER DEALINGS IN THE SOFTWARE.
 *  
 * https://opensource.org/licenses/MIT
 * 
 */
 
 
 
 

// /////////////////////////////////////////////////////////////////
// State-Action Fabric Element 
// SAFE is a micro-container for server-side SAM implementations
// 
// SAFE implements the following services
//  - action dispatcher 
//  - action hang back
//  - session dehydration / hydration
//  - enforces allowed actions and action "hang back"
//  - logging
//
//  TODO
//  - server side time travel
//  - caching (idempotent actions)


'use strict' ;

// import { saveSnapshot, getSnapshot } from './timeTravelStore.js'  ;

let safe = {} ;


// minimal service implementations

// session manager 

safe.defaultSessionManager = {
    
    dehydrateSession: (model) => {
        if (model.__token) {
            safe.defaultSessionManager[model.__token] = model.__session ;
        }
    },
    
    rehydrateSession: (token) => {
        let session =  safe.defaultSessionManager[token] ;
        return session ; 
    }
    
} ;

// logger

safe.defaultLogger = {
    output: (level,message) => {
        var timestamp = new Date().getTime() ;
        var m = message ;
        if (typeof message !== 'string') {
            m = JSON.stringify(message) ;
        }
        console.log([level,timestamp,m].join(' : ')) ;
    },
    warning: (message) => {
        if (safe.logger.loggingLevel <=1) { safe.logger.output('[WARNING] ',message) ; }
    },
    info: (message) => {
        if (safe.logger.loggingLevel <=0) { safe.logger.output('[INFO]    ',message) ; }
    },
    error: (message) => {
        if (safe.logger.loggingLevel <=2) { safe.logger.output('[ERROR]   ',message) ; }
    },
    loggingLevel: 0
} ;

// error handler

safe.defaultErrorHandler = (message) => {
   safe.logger.error(message) ; 
} ;

// default in memory store for time travel 

safe.defaultSnapshotStore = (function() {
    let store = [] ;
    
    return {
        
        store: (i,s) => {
            store[i] = s ;
            return store.length ;
        },
        
        retrieve: (i) => {
            if ((i>=0) && (i<store.length)) {
                
                return store[i] ;
                
            } else {
                
                return {} ;
                
            }
        },
        
        retrieveAll: () => {
            return store ;
        },
        
        length: () => {
            return store.length ;
        }
    } ;
    
})() ;

// time traveler

safe.defaultTimeTraveler = (store) => {
    let snapshotStore = store || safe.defaultSnapshotStore ;
    let cursor = -1 ;
    let display =(res,representation) => {
        res.status(200).send(representation) ;
    } ; 
    return {
        init: (app,path, next) => {
            
            // substitute default response function with next
            display = next || display ;
            
            // API : returns a given snapshot or all if index is negative 
            app.get(path+'/:snapshot', function(req,res) { 
                
                let index = req.params.snapshot ;
                res.setHeader("Content-Type", "application/json");
                display(res,JSON.stringify(safe.getSnapshot(index))) ;
                
            }) ;
            
            app.get(path, function(req,res) { 
                res.setHeader("Content-Type", "application/json") ;
                display(res,JSON.stringify(safe.getSnapshot())) ;
                
            }) ;
            
            // travel back
            app.post(path+'/:snapshot', function(req,res) { 
                
                let dis = (representation) => {
                    let resp = res ;
                    display(resp,representation) ;
                }
                
                let index = req.params.snapshot ;
                
                if (index>=0) {
                    let snapshot = snapshotStore.retrieve(index) ;
                    if ((index>=0) && (index<snapshotStore.length())) {
                        cursor = index ;
                    } 
                    let m = safe.deepCopy(snapshot.store) ;
                    m.__token = req.cookies['safe_token'] || '' ;
                    
                    let modelProperties = Object.getOwnPropertyNames(safe.model).filter(function (p) {
                        return typeof safe.model[p] !== 'function';
                    }) ;
                    
                    let snapShotProperties = Object.getOwnPropertyNames(m) ;
                    
                    modelProperties.forEach(function(p) {
                        delete safe.model[p] ;
                    });
                    
                    snapShotProperties.forEach(function(p) {
                       safe.model[p] = m[p] ; 
                    });
                    
                    safe.render(safe.model,dis,false) ;
                }
            }) ;
        },
        
        saveSnapshot: (model,dataset) => {
            let snapshot = snapshotStore.retrieve(cursor) ;
            if (dataset) {
                cursor++ ;
                snapshot = {} ;
                snapshot.dataset = safe.deepCopy(dataset) ;
            }
            
            if (model) {
                snapshot.store = safe.deepCopy(model) ;
            } 
            
            return snapshotStore.store(cursor,snapshot) ;
            
        },
        
        getSnapshot: (i) => {
            i = i || -1 ;
            
            if (i>=0) { 
                return snapshotStore.retrieve(i); 
            } else {
                return snapshotStore.retrieveAll() ;
            }
        },
        
        displayTimeTravelControls: (representation) => {
            
            return (representation + 
            '          <br>\n<br>\n<hr>\n<div class="mdl-cell mdl-cell--6-col">\n'+
            '                      <input id="__snapshot" type="text" class="form-control"><br>\n'+
            '                      <button id="__travel" onclick="JavaScript:return travel({\'snapshot\':document.getElementById(\'__snapshot\').value});"> TimeTravel </button>\n'+
            '          </div><br><br>\n') ;
        }
        
        
    } ;
} ;


// SAFE Core

// Insert SAFE middleware and wire SAM components

safe.init = (actions, model, state, view, logger, errorHandler, sessionManager) => {
    
    safe.actions = actions ;
    safe.model = model ;
    safe.state = state ;
    safe.view = view ;
    
    safe.errorHandler = errorHandler || safe.defaultErrorHandler ;
    
    safe.logger = logger || safe.defaultLogger ;
    
    safe.hangback = false ;
    safe.steps = [] ;
    safe.stepCount = 0 ;
    safe.lastStep = safe.newStep(safe.stepCount) ;
    
    safe.sessionManager = sessionManager || safe.defaultSessionManager ;
    
    safe.allowedActions = null ;
    
    if (actions) {
        if (model) {
            actions.init(safe.present) ;       
        }
    } 
    
    if (state) {
        
        if (model) {
            model.init(safe.render) ;
            state.init(safe.render, safe.view) ;
        }
        
        if (view) {
            if (actions) {
                state.init(null,view, safe.display, actions.intents) ;
            } else {
                state.init(null,view, safe.display) ;
            }
        }
        
        safe.allowedActions = state.allowedActions || [] ;
    }
} ;

safe.initTimeTraveler = (timeTraveler) => {
    if (timeTraveler) {
        safe.saveSnapshot = timeTraveler.saveSnapshot || null ;
        safe.getSnapshot = timeTraveler.getSnapshot || null ;
        safe.displayTimeTravelControls = timeTraveler.displayTimeTravelControls || null ;
    } 
} ;

// The Dispatcher is an optional component which 
// exposes a dispatch API 

safe.dispatcher = (app,path,next) => {
    // assumes express cookie-parser middleware
    
    app.post(path, function(req,res) { 
        var ret = (representation) => {
                        res.status(200).send(representation) ;
                    } ;
        var data = req.body ;
        
        data.__token = req.cookies['safe_token'] || '' ;
        
        safe.dispatch(data.__action,data,next || ret) ;
    }) ;
    
} ;

// The dispatch method decides whether an action can be dispatched
// based on SAFE's context

safe.dispatch = (action,data,next) => {
    data = data || {} ;
    var dispatch = false ;
    const lastStepActions = safe.lastStep.actions ; 

    safe.logger.info('dispatcher received request'+JSON.stringify(data)) ;
    safe.logger.info('lastStepActions            '+JSON.stringify(lastStepActions)) ;
    if (lastStepActions.length === 0) {
        // action validation is disabled
        dispatch = true ;
    } else {
        lastStepActions.forEach( (a) => {
            safe.logger.info(a) ;
            if (a.action === action) {
                dispatch = true ;
                // tag the action with the stepid
                // we want to enforce one action per step
                // if the step does not match we should not dispatch
                data.__actionId = a.uid ;
                safe.logger.info('tagging action with            '+JSON.stringify(a)) ;
            
                safe.lastStep.dispatched = action ;
            }
        }) ;
    }
    if (!dispatch) {
        safe.errorHandler({action: action, error: "not allowed"}) ;
    } else {
            
        if (safe.actions[action] !== undefined) {
            
            // dispatch action
            safe.logger.info('invoking action            '+JSON.stringify(data)) ;
            safe.actions[action](data,next) ;
           
        } else { 
            safe.errorHandler({action: action, error: "not found"}) ;  
        }
        
    }

} ;


safe.present = (data,next) => {
    const actionId = data.__actionId || null ;
    
    if (!safe.blocked) {
        const lastStepActions = safe.lastStep.actions ; 
        safe.logger.info(lastStepActions) ;
        var presentData = (lastStepActions.length === 0) ;
        if (!presentData) { 
            // are we expecting that action?
            lastStepActions.forEach(function(el) {
                if (el.uid === actionId) {
                    presentData = true ;
                }
            }); 
        }
        if (presentData) {
            safe.block() ;
            if (data.__token) {
                safe.model.__session = safe.sessionManager.rehydrateSession(data.__token) ;
                safe.model.__token = data._token ;
                
            }
            if (safe.saveSnapshot) {
                // Store snapshot in TimeTravel
                safe.saveSnapshot(null,data) ;
            }
            safe.model.present(data,next) ;
        } else {
            // unexpected actions
            // possibly an action from previous step that needs to be blocked
        }
    } else {
        // ignore action's effect
        safe.logger({blocked: true,data}) ;
    }    
} ;

safe.block = () => {
    safe.lastStep.actions = [] ;
    safe.blocked = true ;
} ;

safe.unblock = () => {
    safe.blocked = false ;
} ;

safe.newStep = (uid,allowedActions) => {
    allowedActions = allowedActions || [ ] ;
    var k = 0 ;
    var step = { 
        stepid: uid , actions: allowedActions.map((action) => {
            var auid = uid + '_' + k++ ;
            return { action, uid: auid } ;   
    })  } ;
    
    safe.steps.push(step) ;
    safe.logger.info('new step        :'+JSON.stringify(step)) ;
    return step ;
} ;

safe.render = (model,next, takeSnapShot) => {
    takeSnapShot = takeSnapShot || true ;
    if (takeSnapShot && safe.saveSnapshot) {
        // Store snapshot in TimeTravel
        safe.saveSnapshot(model,null) ;
    }
    safe.allowedActions = safe.state.representation(model,next) || [] ;
    safe.unblock() ;
    safe.lastStep = safe.newStep(safe.stepCount++, safe.allowedActions) ;
    safe.sessionManager.dehydrateSession(model) ;
    safe.state.nextAction(model) ;
} ;

safe.display = (representation,next) => {
    if (safe.displayTimeTravelControls) {
        representation = safe.displayTimeTravelControls(representation) ;
    }
    safe.view.display(representation,next) ;
} ;

safe.deepCopy = function(x) {
    return JSON.parse(JSON.stringify(x)) ;
} ;




module.exports = safe ;

File: SamSafeCSharp/Components/Safe.cs
/******************************************************************************************* 
 * The MIT License (MIT)
 * -----------------------------------------------------------------------------------------
 * Copyright (c) 2016 Convergence Modeling LLC
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of 
 * this software and associated documentation files (the "Software"), to deal in the 
 * Software without restriction, including without limitation the rights to use, copy, 
 * modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, 
 * and to permit persons to whom the Software is furnished to do so, subject to the 
 * following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all 
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR 
 * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE 
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, 
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE 
 * USE OR OTHER DEALINGS IN THE SOFTWARE.
 *  
 * https://opensource.org/licenses/MIT
 * 
 */

using System;
using System.Collections.Generic;
using SamSafeCSharp.Helpers;


// /////////////////////////////////////////////////////////////////
// State-Action Fabric Element 
// SAFE is a micro-container for server-side SAM implementations
// 
// SAFE implements the following services
//  - action dispatcher 
//  - action hang back
//  - session dehydration / hydration
//  - enforces allowed actions and action "hang back"
//  - logging
//
//  TODO
//  - server side time travel
//  - caching (idempotent actions)
// import { saveSnapshot, getSnapshot } from './timeTravelStore.js'  ;





namespace SamSafeCSharp.Components
{
    public class Safe
    {
        private Func<Model, string, int> _saveSnapshot;
        private Actions _actions;
        private Model _model;
        private State _state;
        private View _view;
        private DefaultLogger _logger;
        private Action<string> _errorHandler;
        private DefaultSessionManager _sessionManager;
        private List<Step> _steps;
        private int _stepCount;
        private Step _lastStep;
        private string[] _allowedActions;
        private bool _hangback;
        private Func<int, string> _getSnapshot;
        private Action<PresenterModel, Action<string>> _present;
        private Action<Model, Action<string>> _render;
        private Action<string, Action<string>> _display;
        private bool _blocked;
        private Func<string, string> _displayTimeTravelControls;


        // SAFE Core

        // Insert SAFE middleware and wire SAM components
        public void Init(Actions actions, Model model, State state, View view, DefaultLogger logger = null, Action<string> errorHandler = null, DefaultSessionManager sessionManager = null)
        {
            this._actions = actions;
            this._model = model;
            this._state = state;
            this._view = view;

            this._errorHandler = errorHandler ?? DefaultErrorHandler;
            this._logger = logger ?? new DefaultLogger();

            this._hangback = false;
            this._steps = new List<Step>();
            this._stepCount = 0;
            this._lastStep = this.NewStep(this._stepCount);

            this._sessionManager = sessionManager ?? new DefaultSessionManager();

            this._allowedActions = null;

            if (actions != null && model != null)
            {
                actions.Init(this._present);
            }

            if (state != null)
            {
                if (model != null)
                {
                    model.Init(this.Render);
                    state.Init(this._render, this._view);
                }
            }
            if (view != null)
            {
                if (actions != null)
                {
                    state.Init(null, view, this._display, actions.Intents);
                }
                else
                {
                    state.Init(null, view, this._display);
                }
                this._allowedActions = state.AllowedActions ?? new string[] { };
            }
        }

        public void InitTimeTraveler(DefaultTimeTraveler timeTraveler)
        {
            if (timeTraveler != null)
            {
                this._saveSnapshot = timeTraveler.SaveSnapshot;
                this._getSnapshot = timeTraveler.GetSnapshot;
                this._displayTimeTravelControls = timeTraveler.DisplayTimeTravelControls;
            }
        }


        // The dispatch method decides whether an action can be dispatched
        // based on SAFE's context
        public void Dispatch(string action, PresenterModel data, Action<string> next)
        {
            this._logger.Info("dispatcher received request");
            bool dispatch = false;
            var lastStepActions = this._lastStep.Actions;

            this._logger.Info("dispatcher received request" + JsHelpers.JSON.stringify(data));
            this._logger.Info("lastStepActions            " + JsHelpers.JSON.stringify(lastStepActions));


            if (lastStepActions.Count == 0)
            {
                // action validation is disabled
                dispatch = true;
            }
            else
            {
                foreach (var lastStep in lastStepActions)
                {
                    this._logger.Info(lastStep.ToString());
                    if (lastStep.Action == action)
                    {
                        dispatch = true;
                        // tag the action with the stepid
                        // we want to enforce one action per step
                        // if the step does not match we should not dispatch
                        data.__actionId = lastStep.UId;
                        this._logger.Info("tagging action with            " + lastStep.ToString());

                        this._lastStep.Dispatched = action;
                    }
                }
            }

            if (!dispatch)
            {
                this._errorHandler(new { action = action, error = "not allowed" }.ToString());
            }
            else
            {

                if (this._actions.ActionList.ContainsKey(action))
                {
                    // dispatch action
                    this._logger.Info("invoking action            " + data.ToString());
                    this._actions.ActionList[action](data, next);
                }
                else
                {
                    this._errorHandler(new { action = action, error = "not found" }.ToString());
                }

            }
        }

        public void Present(PresenterModel data, Action<string> next)
        {
            string actionId = data.ActionId ?? null;

            if (!this._blocked)
            {
                var lastStepActions = this._lastStep.Actions;
                this._logger.Info(lastStepActions.ToString());
                bool presentData = (lastStepActions.Count == 0);

                if (!presentData)
                {
                    // are we expecting that action?
                    foreach (var item in lastStepActions)
                    {
                        if (item.UId == actionId)
                        {
                            presentData = true;
                        }
                    }
                }
                if (presentData)
                {
                    Block();
                    if (!string.IsNullOrEmpty(data.__token))
                    {
                        this._model.__session = this._sessionManager.RehydrateSession(data.__token);
                        this._model.__token = data.__token;
                    }
                    if (this._saveSnapshot != null)
                    {
                        // Store snapshot in TimeTravel
                        this._saveSnapshot(null, data.ToString());
                    }
                    this._model.Present(data, next);
                }
                else
                {
                    // unexpected actions
                    // possibly an action from previous step that needs to be blocked
                }
            }
            else
            {
                // ignore action's effect
                // this.logger({ blocked: true,data}) ; //todo
            }
        }

        public void Block()
        {
            this._lastStep.Actions = new List<StepAction>();
            this._blocked = true;
        }

        public void UnBlock()
        {
            this._blocked = false;
        }

        public Step NewStep(int uId = 0, string[] allowedActions = null)
        {
            _allowedActions = allowedActions ?? new string[] { };
            Step step = new Step(uId, allowedActions);

            this._steps.Add(step);
            this._logger.Info("new step        :" + JsHelpers.JSON.stringify(step));
            return step;
        }

        public void Render(Model model, Action<string> next)
        {
            Render(model, next, true);
        }

        public void Render(Model model, Action<string> next, bool takeSnapshot)
        {
            //todo check next lines
            /*takeSnapShot = takeSnapShot || true;
            if (takeSnapShot && this.saveSnapshot)
            {
                // Store snapshot in TimeTravel
                this.saveSnapshot(model, null);
            }*/

            this._allowedActions = this._state.Representation(model, next);
            this.UnBlock();
            this._lastStep = this.NewStep(this._stepCount++, this._allowedActions);
            this._sessionManager.DehydrateSession(model);
            this._state.NextAction(model);
        }

        public void Display(string representation, Action<string> next)
        {

            if (this._displayTimeTravelControls != null)
            {
                representation = this._displayTimeTravelControls(representation);
            }
            this._view.Display(representation, next);

        }

        public void DefaultErrorHandler(string message = "")
        {

            this._logger.Error(message);
        }
    }


    // minimal service implementations

    // session manager 
    // logger

    // default in memory store for time travel 
    public class DefaultSnapshotStore
    {
        readonly string[] _store = new string[] { };

        public int Store(int i, string s)
        {
            _store[i] = s;
            return _store.Length;
        }

        public string Retrieve(int i)
        {
            if (i >= 0 && i < _store.Length)
                return _store[i];
            return null;
        }

        public string[] RetrieveAll()
        {
            return _store;
        }
    }

    // time traveler
    public class Step
    {
        public Step(int uId, string[] allowedActions)
        {
            this.StepId = uId;
            this.Actions = new List<StepAction>();
            if (allowedActions != null)
            {
                var k = 0;
                foreach (var allowedAction in allowedActions)
                {
                    this.Actions.Add(new StepAction(uId + "_" + k++, allowedAction));
                }
            }
        }
        public int StepId { get; set; }
        public List<StepAction> Actions { get; set; }

        public string Dispatched;
    }

    public class StepAction
    {
        public StepAction(string uid, string allowedAction)
        {
            this.UId = uid;
            this.Action = allowedAction;
        }
        public string UId { get; }
        public string Action { get; }

        public override string ToString()
        {
            return $"{UId} - {Action}";
        }
    }
}


